include/deal.II/lac/petsc_ts.templates.h

```diff
@@ -818,25 +1024,78 @@ namespace PETScWrappers
         AssertPETSc(TSGetSNES(ts, &snes));
         AssertPETSc(SNESGetKSP(snes, &ksp));
         AssertPETSc(KSPSetType(ksp, KSPPREONLY));
-        AssertPETSc(KSPSetPC(ksp, precond.get_pc()));
+        AssertPETSc(KSPSetPC(ksp, solve_with_jacobian_pc.get_pc()));
       }
 
+    if (distribute)
+      AssertPETSc(TSSetPostStage(ts, ts_poststage));
+
     // Attach user monitoring routine.
     if (monitor)
       AssertPETSc(TSMonitorSet(ts, ts_monitor, this, nullptr));
 
+    // Handle AMR.
+    if (prepare_for_coarsening_and_refinement)
+      {
+        AssertThrow(interpolate,
+                    StandardExceptions::ExcFunctionNotProvided("interpolate"));
+#  if DEAL_II_PETSC_VERSION_GTE(3, 20, 0)
+        (void)ts_poststep_amr;
+        AssertPETSc(TSSetResize(ts, ts_resize_setup, ts_resize_transfer, this));
+#  else
+        AssertPETSc(PetscObjectComposeFunction(
+          (PetscObject)ts,
+          "__dealii_ts_resize_setup__",
+          static_cast<PetscErrorCode (*)(
+            TS, PetscInt, PetscReal, Vec, PetscBool *, void *)>(
+            ts_resize_setup)));
+        AssertPETSc(PetscObjectComposeFunction(
+          (PetscObject)ts,
+          "__dealii_ts_resize_transfer__",
+          static_cast<PetscErrorCode (*)(TS, PetscInt, Vec[], Vec[], void *)>(
+            ts_resize_transfer)));
+        AssertPETSc(TSSetPostStep(ts, ts_poststep_amr));
+#  endif
+      }
+
     // Allow command line customization.
     AssertPETSc(TSSetFromOptions(ts));
 
-    // Handle algebraic components.
+    // By default PETSc does not check for Jacobian errors in optimized
+    // mode. Here we do it unconditionally.
+#  if DEAL_II_PETSC_VERSION_GTE(3, 11, 0)
+    {
+      SNES snes;
+      AssertPETSc(TSGetSNES(ts, &snes));
+      AssertPETSc(SNESSetCheckJacobianDomainError(snes, PETSC_TRUE));
+    }
+#  endif
+  }
+
+
+
+  template <typename VectorType, typename PMatrixType, typename AMatrixType>
+  DEAL_II_CXX20_REQUIRES(
+    (concepts::is_dealii_petsc_vector_type<VectorType> ||
+     std::constructible_from<
+       VectorType,
+       Vec>)&&(concepts::is_dealii_petsc_matrix_type<PMatrixType> ||
+               std::constructible_from<
+                 PMatrixType,
+                 Mat>)&&(concepts::is_dealii_petsc_matrix_type<AMatrixType> ||
+                         std::constructible_from<AMatrixType, Mat>))
+  void TimeStepper<VectorType, PMatrixType, AMatrixType>::
+    setup_algebraic_constraints(const VectorType &y)
+  {
```

Comment:
I can't seem to comment on the lines in question below (line 1104 in the changed version). This is calling `algebraic_components()`, which while not marked as a user-callback, is actually one. It should be wrapped in the `call_and_...()` thingy, and then you don't need to catch exceptions where you call *this* function.